modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java

```diff
@@ -0,0 +1,410 @@
+
+    /**
+     * Constructor.
+     *
+     * @param cur Asynchronous query cursor.
+     */
+    public AsyncResultSetImpl(AsyncSqlCursor<List<Object>> cur, BatchedResult<List<Object>> page, int pageSize, Runnable closeRun) {
+        this.cur = cur;
+        this.batchPage = page;
+        this.pageSize = pageSize;
+        this.closeRun = closeRun;
+        this.page = new Page();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable ResultSetMetadata metadata() {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasRowSet() {
+        return cur.queryType() == SqlQueryType.QUERY || cur.queryType() == SqlQueryType.EXPLAIN;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long affectedRows() {
+        if (hasRowSet() || cur.queryType() == SqlQueryType.DDL) {
+            return -1;
+        }
+
+        assert batchPage.items().size() == 1
+                && batchPage.items().get(0).size() == 1
+                && batchPage.items().get(0).get(0) instanceof Long
+                && !batchPage.hasMore() : "Invalid DML result: " + batchPage;
+
+        return (long) batchPage.items().get(0).get(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean wasApplied() {
+        if (hasRowSet() || cur.queryType() == SqlQueryType.DML) {
+            return false;
+        }
+
+        assert batchPage.items().size() == 1
+                && batchPage.items().get(0).size() == 1
+                && batchPage.items().get(0).get(0) instanceof Boolean
+                && !batchPage.hasMore() : "Invalid DDL result: " + batchPage;
+
+        return (boolean) batchPage.items().get(0).get(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Iterable<SqlRow> currentPage() {
+        if (!hasRowSet()) {
+            throw new NoRowSetExpectedException("Query hasn't result set: [type=" + cur.queryType() + ']');
+        }
+
+        return page;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<? extends AsyncResultSet> fetchNextPage() {
+        if (next == null) {
+            synchronized (mux) {
+                if (next == null) {
+                    if (!hasMorePages()) {
+                        next = HAS_NO_MORE_PAGE_FUTURE;
+                    } else {
+                        next = cur.requestNextAsync(pageSize)
+                                .thenApply(batchRes -> new AsyncResultSetImpl(cur, batchRes, pageSize, closeRun));
+                    }
+                }
+
+            }
+        }
+
+        return next;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasMorePages() {
+        return batchPage.hasMore();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<Void> closeAsync() {
+        return cur.closeAsync().thenAccept((v) -> closeRun.run());
+    }
+
+    private class Page implements Iterable<SqlRow> {
```

Comment:
Can the Page be static class?
Does it need an implicit reference to the parent object?